### DISCUSSION TITLE
[VSPHERE] [IPI]  Upgrade for 4.15 FCOS to 4.16 SCOS worker nodes are not updated to non-secure boot

### DISCUSSION THREAD

**Discussion by fortinj66** [okd-project/okd#2049] [Pre-Release Testing]
After doing a new 4.15 FCOS IPI install successfully and doing the workaround in https://github.com/okd-project/okd-web/pull/22/files I had some slight issues:

- **Worker nodes were not updated to disable secure boot setting**.  Once secure boot was disabled worker nodes updated successfully.  _**Control-plane nodes had no issues**_.

- **New worker nodes after upgrade still had secure boot enabled**.  Once secure boot was disabled worker was created successfully.

Other than the kind of ugly workaround above and the secure boot issues the upgrade seemed clean...  I have a lot more testing to do as we use a lot of add-ins like Ceph and Istio and I need to see how those upgrades do.  I have worries about Kernel versions dependencies.

4.15 version: 4.15.0-0.okd-2024-02-10-035534 (Stable)
4.16 version: 4.16.0-0.okd-scos-2024-10-28-025401 (Nightly but mirrored)

As a side note, a new install for 4.16.0-0.okd-scos-2024-10-28-025401 worked fine for both control-plane and worker nodes.  I did not test adding a new worker node.

**Comment by fortinj66** [comment]
Update: I installed a new 4.16 cluster (same version) and I was able to add a new worker node with no issues.  So something in the upgrade process is not being updated correctly to disable secure boot on  new worker nodes

**Comment by jcpowermac** [comment]
https://github.com/openshift/installer/pull/8793

its only set in the installer to disable secureboot if enabled, for an upgrade you would need to change existing template

**Reply by fortinj66** [reply]
That doesn't explain why control-plane nodes were upgraded without any intervention.  Only the worker nodes had issues.

On large clusters this will be a significant issue with upgrades.

**Reply by jcpowermac** [reply]
Did the control plane nodes actually pivot into scos?

**Reply by fortinj66** [reply]
Actually, I'll take that back.  

After re-installing a new 4.15 FCOS cluster, the control-plane nodes do NOT have secure boot enabled while the worker nodes do.  

That being the case, the upgrade process, in my opinion, should check if secure boot is enabled and disable it...  as should the add node process

**Reply by fortinj66** [reply]
and yes, they went to SCOS with no issues...

**Reply by jcpowermac** [reply]
4.15 we were still using terraform

https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#boot-options
> [efi_secure_boot_enabled](https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#efi_secure_boot_enabled) - (Optional) Use this option to enable EFI secure boot when the firmware type is set to is efi. Default: false.

4.16 uses capi,capv

The only solution for 4.15 clusters wanting to upgrade to 4.16 scos will be to update the template disabling secure boot

**Reply by fortinj66** [reply]
capi,capv doesn't have the capability to disable secure boot on vm creation?

**Reply by fortinj66** [reply]
So really, what is needed is a tool that does all the "pre-upgrade" tasks that need to be done prior to the upgrade.

- change template to disable secure boot
- shutdown each node with secure node enabled, disable it and restart.  (with proper cordon/drain process)

Will the workaround described in https://github.com/okd-project/okd-web/pull/22/files be fixed prior to a stable 4.16 release? or is this not fixable

**Reply by jcpowermac** [reply]
> capi,capv doesn't have the capability to disable secure boot on vm creation?

Nope, which is why I implemented it the way I did. This is only an issue in OKD and FCOS, OCP and RHCOS do not have secureboot enabled.

> So really, what is needed is a tool that does all the "pre-upgrade" tasks that need to be done prior to the upgrade.
> 
>     * change template to disable secure boot
> 
>     * shutdown each node with secure node enabled, disable it and restart.  (with proper cordon/drain process)
> 
> 

Sounds like that would be fairly trivial with powercli or govc, do you want to give that a shot?

> Will the workaround described in https://github.com/okd-project/okd-web/pull/22/files be fixed prior to a stable 4.16 release? or is this not fixable

You may want to ask this on the PR you linked.

**Reply by fortinj66** [reply]
> > So really, what is needed is a tool that does all the "pre-upgrade" tasks that need to be done prior to the upgrade.
> > ```
> > * change template to disable secure boot
> > 
> > * shutdown each node with secure node enabled, disable it and restart.  (with proper cordon/drain process)
> > ```
> 
> Sounds like that would be fairly trivial with powercli or govc, do you want to give that a shot?
> 

I'll take a look at at...  The complexity is the cordon/drain/shutdown processing... probably need to do it by machine sets and make sure we follow the max number of nodes down constraints.

> > Will the workaround described in https://github.com/okd-project/okd-web/pull/22/files be fixed prior to a stable 4.16 release? or is this not fixable
> 
> You may want to ask this on the PR you linked.

right...

**Comment by fortinj66** [comment]
Initial stab at an upgrade script.
https://github.com/fortinj66/okd-4.15-upgrade-scripts
